dali/util/custream.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/util/custream.h"
+#include "dali/util/cuda_utils.h"
+
+namespace dali {
+
+CUStream::CUStream(int device_id, bool default_stream) : created_{false}, stream_{0} {
+  if (!default_stream) {
+    int orig_device;
+    cudaGetDevice(&orig_device);
```

Comment:
`cudaSetDevice - This call may be made from any host thread, to any device, and at any time. This function will do no synchronization with the previous or new device, and should be considered a very low overhead call. `
```
if device_id >= 0:
  cudaSetDevice(device_id);
```
Or even you can use DeviceGuard.

Reply:
Done